pkg/kubectl/cmd/apply.go

```diff
@@ -0,0 +1,161 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/util/strategicpatch"
+)
+
+// ApplyOptions stores cmd.Flag values for apply.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type ApplyOptions struct {
+	Filenames []string
+}
+
+const (
+	apply_long = `Apply a configuration to a resource by filename or stdin.
+
+JSON and YAML formats are accepted.`
+	apply_example = `# Apply the configuration in pod.json to a pod.
+$ kubectl apply -f ./pod.json
+
+# Apply the JSON passed into stdin to a pod.
+$ cat pod.json | kubectl apply -f -`
+)
+
+func NewCmdApply(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &ApplyOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "apply -f FILENAME",
+		Short:   "Apply a configuration to a resource by filename or stdin",
+		Long:    apply_long,
+		Example: apply_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(validateArgs(cmd, args))
+			cmdutil.CheckErr(cmdutil.ValidateOutputArgs(cmd))
+			cmdutil.CheckErr(RunApply(f, cmd, out, options))
+		},
+	}
+
+	usage := "Filename, directory, or URL to file that contains the configuration to apply"
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+	cmd.MarkFlagRequired("filename")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddOutputFlagsForMutation(cmd)
+	return cmd
+}
+
+func validateArgs(cmd *cobra.Command, args []string) error {
+	if len(args) != 0 {
+		return cmdutil.UsageError(cmd, "Unexpected args: %v", args)
+	}
+
+	return nil
+}
+
+func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *ApplyOptions) error {
+	shortOutput := cmdutil.GetFlagString(cmd, "output") == "name"
+	schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
+	if err != nil {
+		return err
+	}
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object()
+	r := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		Schema(schema).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options.Filenames...).
+		Flatten().
+		Latest().
```

Comment:
Remove Latest() here. The reason is that the resource builder will overwrite `info.Object` when you call `Visit` with the latest version from server. You want to save/serialize the object as loaded from file *before* you get the latest version of the object in order to do a 3-way patch. See suggested changes below.